### ISSUE TITLE
Scylla Ubuntu meta package doesn't specify versions of its dependencies

### ISSUE THREAD

**Issue by mikejquinn** [scylladb/scylla#5514]
*Installation details*
Scylla version (or git commit hash): 3.1.0-0.20191012.9c3cdded9-1
Cluster size: N/A
OS (RHEL/CentOS/Ubuntu/AWS AMI): Ubuntu 18.04

*Problem*
Installing the 3.1.0 version of Scylla using the `scylla` metapackage will install more recent 3.1.x versions of scylla.

First, use `apt` to install version 3.1.0 of scylla:
```
ubuntu@dev-mike-scylla001.us-e:/etc/apt/sources.list.d$ sudo apt install scylla=3.1.0-0.20191012.9c3cdded9-1
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following additional packages will be installed:
  scylla-conf scylla-jmx scylla-kernel-conf scylla-python3 scylla-server scylla-tools scylla-tools-core
The following NEW packages will be installed:
  scylla scylla-conf scylla-jmx scylla-kernel-conf scylla-python3 scylla-server scylla-tools scylla-tools-core
```
`apt` will then install the most recent 3.1.x version of Scylla:
```
Preparing to unpack .../0-scylla-conf_3.1.2-0.20191127.9d78d848e-1_amd64.deb ...
Unpacking scylla-conf (3.1.2-0.20191127.9d78d848e-1) ...
Selecting previously unselected package scylla-python3.
Preparing to unpack .../1-scylla-python3_3.7.2-0.20191127.9d78d848e-1_amd64.deb ...
Unpacking scylla-python3 (3.7.2-0.20191127.9d78d848e-1) ...
Selecting previously unselected package scylla-server.
Preparing to unpack .../2-scylla-server_3.1.2-0.20191127.9d78d848e-1_amd64.deb ...
Unpacking scylla-server (3.1.2-0.20191127.9d78d848e-1) ...
Selecting previously unselected package scylla-jmx.
...
...
```

This was a problem for us because we spun up a new Scylla node which then started alerting because it was on a different version from the rest of the fleet. It seems like the meta package should be declaring the correct versions of its dependencies as 3.1.0 here.

**Comment by syuu1228** [comment]
This is our packaging / versioning scheme issue.
On .rpm version of scylla metapackage we have dependency to specific version of scylla packages, like this:
```
Requires:       scylla-server = 3.1.0 scylla-jmx = 3.1.0 scylla-tools = 3.1.0 scylla-tools-core = 3.1.0 scylla-kernel-conf = 3.1.0
```
Then following command can install specified version of scylla packages:
```
# yum install scylla-3.1.0
```
However, Debian packaging system is different.
When comparing version number with equal (=), Debian packaging system requires full version number including revision number.
So "Depends" field of meta package become like this:
```
 Depends: scylla-server (= 3.1.0-0.20191012.9c3cdded9-1), scylla-kernel-conf (= 3.1.0-0.20191012.9c3cdded9-1), scylla-conf (= 3.1.0-0.20191012.9c3cdded9-1), scylla-jmx (= 3.1.0-20191012.d17d5bc-1), scylla-tools (= 3.1.0-20191012.132fdf208d-1), scylla-tools-core (= 3.1.0-20191012.132fdf208d-1)

# Note that we can use "(= ${binary:Version})" for scylla main repo packages
```
Problem is, our versioning scheme contains git hash, and scylla-server / jmx / tools has different hash.
Since scylla metapackage is part of scylla-server repo, deb build script does not able to know jmx / tools git hash.

To solve this issue, we need to pass git hash of all repositories.
I think it's better to separate off metapackage as single package, move to new git repository that contains all modules as submodule.
And we probably can use "supermodule" repository for it (see: https://docs.google.com/document/d/1XVwTpxYkPcj08_3rJxH_9BQ0IjMjpr9iqat8x8ZmEMs)

But, as a workaround, we might able to just add new build options to specify jmx / tools full version number on build_deb.sh, something like this:
```
./reloc/build_deb.sh --jmx-version <jmx full version number> --tools-version <tools full version number>
```


What do you think, @avikivity ?

**Comment by syuu1228** [comment]
@avikivity ping

**Comment by syuu1228** [comment]
@avikivity ping

**Comment by avikivity** [comment]
We can start by depending on the version part of the version-release string, so requires scylla-kernel-conf (=3.1.0).

Longer term it is better to use the full release too, but that can be done after we have the supermodule.

**Comment by syuu1228** [comment]
@avikivity On .deb, "Depends: scylla-kernel-conf (=3.1.0)" does not matches scylla-kernel-conf-3.1.0-0.20191012.9c3cdded9-1, unfortunately.
And that's the reason why I haven't specified version on "Depends" tag.

Debian packaging system requires to FULL string to match version:
```
[syuu@monster scylla.3]$ dpkg --compare-versions 3.1.0 eq 3.1.0-0.20191012.9c3cdded9-1
[syuu@monster scylla.3]$ echo $?
1
[syuu@monster scylla.3]$ dpkg --compare-versions 3.1.0-0.20191012.9c3cdded9-1 eq 3.1.0-0.20191012.9c3cdded9-1
[syuu@monster scylla.3]$ echo $?
0
```

There is a variable to specify current package version in full string, like this:
"scylla-kernel-conf (= ${binary:Version})"
But it won't match with scylla-jmx and scylla-tools, since we have different revision number on these packages (such as git hash).

reference: https://aalvarez.me/posts/debian-package-versioning/

**Comment by avikivity** [comment]
Let's pass it as a parameter then, and change the build script to pass this parameter.

**Comment by syuu1228** [comment]
@avikivity Implemented and tested it, but installing specific version of packages from metapackage dependency didn't work correctly.
apt just tries to install newest packages, then conflict with metapackage dependency.
```
root@10038aae7e4a:/home/syuu# apt-cache policy scylla-jmx
scylla-jmx:
  Installed: (none)
  Candidate: 666.development-20191218.771cf6e-1
  Version table:
     666.development-20191218.771cf6e-1 500
        500 http://10038aae7e4a:8080 bionic/main amd64 Packages
          666.development-20190904.6bac4c1-1 500
        500 http://10038aae7e4a:8080 bionic/main amd64 Packages
          666.development-20190903.6bac4c1-1 500
        500 http://10038aae7e4a:8080 bionic/main amd64 Packages  
root@10038aae7e4a:/home/syuu# apt-get install scylla=666.development-0.20200131.985e00efa-1
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable  distribution that some required packages have not yet been created  or been moved out of Incoming.
The following information may help to resolve the situation:        
The following packages have unmet dependencies:
 scylla : Depends: scylla-jmx (= 666.development-20190903.6bac4c1-1) but 666.development-20191218.771cf6e-1 is to be installed
E: Unable to correct problems, you have held broken packages.
```
I found Q&A about "resolving a dependency on a fixed version", seems like the answer is it just not possible:
https://unix.stackexchange.com/questions/350192/apt-get-not-properly-resolving-a-dependency-on-a-fixed-version-in-a-debian-ubunt

There is several way to force using older package, but non of them are automatically, requires extra configurations / non standard procedures.
In such condition, I think specifying version for each packages is simplest way (we use it for downgrade).
Something like:
```
sudo apt-get install scylla=3.x.y\* scylla-server=3.x.y\* scylla-jmx=3.x.y\* scylla-tools=3.x.y\* scylla-tools-core=3.x.y\* scylla-kernel-conf=3.x.y\* scylla-conf=3.x.y\*
```

**Comment by eliransin** [comment]
@syuu1228 I have 2 questions:
1. What about using >= and << in depends to eliminate the need for knowing the commit hash?
Something like  >=3.x.0 <<3.x+1.0
2. What about trying to install using aptitude? (maybe even coming up with an automated way to choose the option of downgrade as aptitude is automatic.

**Comment by eliransin** [comment]
Another option is to put in the repo a curl script that does the call that you mentioned above and just 
run it using Curl. Or include the exact command line in the release-notes/configuration.

**Comment by eliransin** [comment]
Did you try the `apt-get` command with --allow-downgrades ?

**Comment by eliransin** [comment]
@syuu1228 ping, see my questions above.

**Comment by eliransin** [comment]
@syuu1228 ping

**Comment by eliransin** [comment]
@avikivity / @syuu1228 I have a working solution for this problem, however it is a little bit cumbersome and the direction needs your approval:
It includes two stages: Add range dependencies in the scylla meta-package and then use aptitude to install those dependencies:
I tested it on enterprise so I will keep the examples using an enterprise version names:
Specifically lets assume that the enterprise version you want to install is 2019.1.11 and there is already a new 2019.1.12 out.

The `Depends` field for the metapackage `scylla-enterpris` should follow the pattern for every dependency package A:
`Depends: A (>= {{version}}), A (<< {{next-package}})`
While in our example it will translate to:
`Depends: A (>= 2019.1.11), A (<< 2019.1.12)`
This alone will not help but at leas will start preventing the wrong version from being installed:

```
apt install "scylla-enterprise=2019.1.11-0.20200810.966418404-0ubuntu1~xenial"
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 scylla-enterprise : Depends: scylla-enterprise-server (< 2019.1.12) but 2019.1.12-0.20200729.fc61a884b-0ubuntu1~xenial is to be installed
                     Depends: scylla-enterprise-jmx (< 2019.1.12) but 2019.1.12-20200729.1c9d551-0ubuntu1~xenial is to be installed
                     Depends: scylla-enterprise-tools (< 2019.1.12) but 2019.1.12-20200729.b348d6d1ea-0ubuntu1~xenial is to be installed
                     Depends: scylla-enterprise-tools-core (< 2019.1.12) but 2019.1.12-20200729.b348d6d1ea-0ubuntu1~xenial is to be installed
                     Depends: scylla-enterprise-kernel-conf (< 2019.1.12) but 2019.1.12-0.20200729.fc61a884b-0ubuntu1~xenial is to be installed
E: Unable to correct problems, you have held broken packages.

```

This is because apt is short sigted :slightly_smiling_face: it will always try to install the latest version of the dependencies.
We can use aptitude to work around that `sudo apt install aptitude`:
Running the following:
```
sudo aptitude update 
sudo aptitude install -V "scylla-enterprise=2019.1.11-0.20200810.966418404-0ubuntu1~xenial"
```
Will result in in the snippet below:
Things to notice is that you have to actively or should I say interactively choose the solution you want.
```
The following NEW packages will be installed:
  ca-certificates-java{a} [20160321ubuntu1]  curl{a} [7.47.0-1ubuntu2.15]  dh-python{a} [2.20151103ubuntu1.2]  file{a} [1:5.25-2ubuntu1.4]  fontconfig-config{a} [2.11.94-0ubuntu1.1]  fonts-dejavu-core{a} [2.35-1]  
  hugepages{a} [2.19-0ubuntu1]  hwloc-nox{a} [1.11.2-3]  java-common{a} [0.56ubuntu2]  libavahi-client3{a} [0.6.32~rc+dfsg-1ubuntu2.3]  libavahi-common-data{a} [0.6.32~rc+dfsg-1ubuntu2.3]  
  libavahi-common3{a} [0.6.32~rc+dfsg-1ubuntu2.3]  libbsd0{a} [0.8.2-1ubuntu0.1]  libcrypto++9v5{a} [5.6.1-9ubuntu0.1]  libcups2{a} [2.1.3-4ubuntu0.11]  libedit2{a} [3.1-20150325-1ubuntu2]  libfontconfig1{a} [2.11.94-0ubuntu1.1]  
  libfreetype6{a} [2.6.1-0.1ubuntu2.4]  libgnutlsxx28{a} [3.4.10-4ubuntu1.8]  libhugetlbfs0{a} [2.19-0ubuntu1]  libhwloc-plugins{a} [1.11.2-3]  libhwloc5{a} [1.11.2-3]  libicu55{a} [55.1-7ubuntu0.5]  
  libjpeg-turbo8{a} [1.4.2-0ubuntu3.4]  libjpeg8{a} [8c-2ubuntu8]  libjsoncpp1{a} [1.7.2-1]  liblcms2-2{a} [2.6-3ubuntu2.1]  libltdl7{a} [2.4.6-0.1]  libmagic1{a} [1:5.25-2ubuntu1.4]  libmpdec2{a} [2.4.2-1]  
  libnspr4{a} [2:4.13.1-0ubuntu0.16.04.1]  libnss3{a} [2:3.28.4-0ubuntu0.16.04.13]  libnss3-nssdb{a} [2:3.28.4-0ubuntu0.16.04.13]  libnuma1{a} [2.0.11-1ubuntu1.1]  libopts25{a} [1:5.18.7-3]  libpci3{a} [1:3.3.1-1.1ubuntu1.3]  
  libpciaccess0{a} [0.13.4-1]  libpcsclite1{a} [1.8.14-1ubuntu1.16.04.1]  libpng12-0{a} [1.2.54-1ubuntu1.1]  libprotobuf9v5{a} [2.6.1-1.3]  libpython-stdlib{a} [2.7.12-1~16.04]  libpython2.7-minimal{a} [2.7.12-1ubuntu0~16.04.12]  
  libpython2.7-stdlib{a} [2.7.12-1ubuntu0~16.04.12]  libpython3-stdlib{a} [3.5.1-3]  libpython3.5-minimal{a} [3.5.2-2ubuntu0~16.04.11]  libpython3.5-stdlib{a} [3.5.2-2ubuntu0~16.04.11]  libsnappy1v5{a} [1.1.3-2]  
  libx11-6{a} [2:1.6.3-1ubuntu2.1]  libx11-data{a} [2:1.6.3-1ubuntu2.1]  libxau6{a} [1:1.0.8-1]  libxcb1{a} [1.11.1-1ubuntu1]  libxdmcp6{a} [1:1.1.2-1.1]  libxext6{a} [2:1.3.3-1]  libxi6{a} [2:1.7.6-1]  
  libxml2{a} [2.9.3+dfsg1-1ubuntu0.7]  libxrender1{a} [1:0.9.9-0ubuntu1]  libxtst6{a} [2:1.2.2-1]  libyaml-0-2{a} [0.1.6-3]  mime-support{a} [3.59ubuntu1]  ntp{a} [1:4.2.8p4+dfsg-3ubuntu5.10]  ocl-icd-libopencl1{a} [2.2.8-1]  
  openjdk-8-jre-headless{a} [8u265-b01-0ubuntu2~16.04]  pciutils{a} [1:3.3.1-1.1ubuntu1.3]  python{a} [2.7.12-1~16.04]  python-cffi-backend{a} [1.5.2-1ubuntu1]  python-chardet{a} [2.3.0-2]  python-cryptography{a} [1.2.3-1ubuntu0.2]  
  python-enum34{a} [1.1.2-1]  python-idna{a} [2.0-3]  python-ipaddress{a} [1.0.16-1]  python-minimal{a} [2.7.12-1~16.04]  python-ndg-httpsclient{a} [0.4.0-3]  python-openssl{a} [0.15.1-2ubuntu0.2]  python-pkg-resources{a} [20.7.0-1]  
  python-pyasn1{a} [0.1.9-1]  python-requests{a} [2.9.1-3ubuntu0.1]  python-six{a} [1.10.0-3]  python-urllib3{a} [1.13.1-2ubuntu0.16.04.3]  python-urwid{a} [1.3.1-2build1]  python-yaml{a} [3.11-3build1]  
  python2.7{a} [2.7.12-1ubuntu0~16.04.12]  python2.7-minimal{a} [2.7.12-1ubuntu0~16.04.12]  python3{a} [3.5.1-3]  python3-minimal{a} [3.5.1-3]  python3-psutil{a} [3.4.2-1ubuntu0.1]  python3-pyudev{a} [0.16.1-3]  
  python3-yaml{a} [3.11-3build1]  python3.5{a} [3.5.2-2ubuntu0~16.04.11]  python3.5-minimal{a} [3.5.2-2ubuntu0~16.04.11]  realpath{a} [8.25-2ubuntu3~16.04]  scylla-enterprise{b} [2019.1.11-0.20200810.966418404-0ubuntu1~xenial]  
  scylla-enterprise-conf{a} [2019.1.12-0.20200729.fc61a884b-0ubuntu1~xenial]  scylla-enterprise-jmx{a} [2019.1.12-20200729.1c9d551-0ubuntu1~xenial]  scylla-enterprise-kernel-conf{a} [2019.1.12-0.20200729.fc61a884b-0ubuntu1~xenial]  
  scylla-enterprise-server{a} [2019.1.12-0.20200729.fc61a884b-0ubuntu1~xenial]  scylla-enterprise-tools{a} [2019.1.12-20200729.b348d6d1ea-0ubuntu1~xenial]  
  scylla-enterprise-tools-core{a} [2019.1.12-20200729.b348d6d1ea-0ubuntu1~xenial]  scylla-gcc73-libatomic1{a} [7.3.0-3ubuntu2~xenialppa1]  scylla-gcc73-libstdc++6{a} [7.3.0-3ubuntu2~xenialppa1]  sgml-base{a} [1.26+nmu4ubuntu1]  
  ucf{a} [3.0036]  x11-common{a} [1:7.7+13ubuntu3.1]  xml-core{a} [0.13+nmu2]  
0 packages upgraded, 103 newly installed, 0 to remove and 2 not upgraded.
Need to get 111 MB of archives. After unpacking 363 MB will be used.
The following packages have unmet dependencies:
 scylla-enterprise : Depends: scylla-enterprise-server (< 2019.1.12) but 2019.1.12-0.20200729.fc61a884b-0ubuntu1~xenial is to be installed.
                     Depends: scylla-enterprise-jmx (< 2019.1.12) but 2019.1.12-20200729.1c9d551-0ubuntu1~xenial is to be installed.
                     Depends: scylla-enterprise-tools (< 2019.1.12) but 2019.1.12-20200729.b348d6d1ea-0ubuntu1~xenial is to be installed.
                     Depends: scylla-enterprise-tools-core (< 2019.1.12) but 2019.1.12-20200729.b348d6d1ea-0ubuntu1~xenial is to be installed.
                     Depends: scylla-enterprise-kernel-conf (< 2019.1.12) but 2019.1.12-0.20200729.fc61a884b-0ubuntu1~xenial is to be installed.
The following actions will resolve these dependencies:

     Keep the following packages at their current version:
1)     scylla-enterprise [Not Installed]                  



Accept this solution? [Y/n/q/?] n
The following actions will resolve these dependencies:

     Install the following packages:                                                          
1)     scylla-enterprise-jmx [2019.1.11-20200727.690bbe5-0ubuntu1~xenial (xenial)]            
2)     scylla-enterprise-kernel-conf [2019.1.11-0.20200727.c98c0150d-0ubuntu1~xenial (xenial)]
3)     scylla-enterprise-server [2019.1.11-0.20200727.c98c0150d-0ubuntu1~xenial (xenial)]     
4)     scylla-enterprise-tools [2019.1.11-20200727.22cc383b70-0ubuntu1~xenial (xenial)]       
5)     scylla-enterprise-tools-core [2019.1.11-20200727.22cc383b70-0ubuntu1~xenial (xenial)]  



Accept this solution? [Y/n/q/?] y
The following NEW packages will be installed:
  ca-certificates-java{a} [20160321ubuntu1]  curl{a} [7.47.0-1ubuntu2.15]  dh-python{a} [2.20151103ubuntu1.2]  file{a} [1:5.25-2ubuntu1.4]  fontconfig-config{a} [2.11.94-0ubuntu1.1]  fonts-dejavu-core{a} [2.35-1]  
  hugepages{a} [2.19-0ubuntu1]  hwloc-nox{a} [1.11.2-3]  java-common{a} [0.56ubuntu2]  libavahi-client3{a} [0.6.32~rc+dfsg-1ubuntu2.3]  libavahi-common-data{a} [0.6.32~rc+dfsg-1ubuntu2.3]  
  libavahi-common3{a} [0.6.32~rc+dfsg-1ubuntu2.3]  libbsd0{a} [0.8.2-1ubuntu0.1]  libcrypto++9v5{a} [5.6.1-9ubuntu0.1]  libcups2{a} [2.1.3-4ubuntu0.11]  libedit2{a} [3.1-20150325-1ubuntu2]  libfontconfig1{a} [2.11.94-0ubuntu1.1]  
  libfreetype6{a} [2.6.1-0.1ubuntu2.4]  libgnutlsxx28{a} [3.4.10-4ubuntu1.8]  libhugetlbfs0{a} [2.19-0ubuntu1]  libhwloc-plugins{a} [1.11.2-3]  libhwloc5{a} [1.11.2-3]  libicu55{a} [55.1-7ubuntu0.5]  
  libjpeg-turbo8{a} [1.4.2-0ubuntu3.4]  libjpeg8{a} [8c-2ubuntu8]  libjsoncpp1{a} [1.7.2-1]  liblcms2-2{a} [2.6-3ubuntu2.1]  libltdl7{a} [2.4.6-0.1]  libmagic1{a} [1:5.25-2ubuntu1.4]  libmpdec2{a} [2.4.2-1]  
  libnspr4{a} [2:4.13.1-0ubuntu0.16.04.1]  libnss3{a} [2:3.28.4-0ubuntu0.16.04.13]  libnss3-nssdb{a} [2:3.28.4-0ubuntu0.16.04.13]  libnuma1{a} [2.0.11-1ubuntu1.1]  libopts25{a} [1:5.18.7-3]  libpci3{a} [1:3.3.1-1.1ubuntu1.3]  
  libpciaccess0{a} [0.13.4-1]  libpcsclite1{a} [1.8.14-1ubuntu1.16.04.1]  libpng12-0{a} [1.2.54-1ubuntu1.1]  libprotobuf9v5{a} [2.6.1-1.3]  libpython-stdlib{a} [2.7.12-1~16.04]  libpython2.7-minimal{a} [2.7.12-1ubuntu0~16.04.12]  
  libpython2.7-stdlib{a} [2.7.12-1ubuntu0~16.04.12]  libpython3-stdlib{a} [3.5.1-3]  libpython3.5-minimal{a} [3.5.2-2ubuntu0~16.04.11]  libpython3.5-stdlib{a} [3.5.2-2ubuntu0~16.04.11]  libsnappy1v5{a} [1.1.3-2]  
  libx11-6{a} [2:1.6.3-1ubuntu2.1]  libx11-data{a} [2:1.6.3-1ubuntu2.1]  libxau6{a} [1:1.0.8-1]  libxcb1{a} [1.11.1-1ubuntu1]  libxdmcp6{a} [1:1.1.2-1.1]  libxext6{a} [2:1.3.3-1]  libxi6{a} [2:1.7.6-1]  
  libxml2{a} [2.9.3+dfsg1-1ubuntu0.7]  libxrender1{a} [1:0.9.9-0ubuntu1]  libxtst6{a} [2:1.2.2-1]  libyaml-0-2{a} [0.1.6-3]  mime-support{a} [3.59ubuntu1]  ntp{a} [1:4.2.8p4+dfsg-3ubuntu5.10]  ocl-icd-libopencl1{a} [2.2.8-1]  
  openjdk-8-jre-headless{a} [8u265-b01-0ubuntu2~16.04]  pciutils{a} [1:3.3.1-1.1ubuntu1.3]  python{a} [2.7.12-1~16.04]  python-cffi-backend{a} [1.5.2-1ubuntu1]  python-chardet{a} [2.3.0-2]  python-cryptography{a} [1.2.3-1ubuntu0.2]  
  python-enum34{a} [1.1.2-1]  python-idna{a} [2.0-3]  python-ipaddress{a} [1.0.16-1]  python-minimal{a} [2.7.12-1~16.04]  python-ndg-httpsclient{a} [0.4.0-3]  python-openssl{a} [0.15.1-2ubuntu0.2]  python-pkg-resources{a} [20.7.0-1]  
  python-pyasn1{a} [0.1.9-1]  python-requests{a} [2.9.1-3ubuntu0.1]  python-six{a} [1.10.0-3]  python-urllib3{a} [1.13.1-2ubuntu0.16.04.3]  python-urwid{a} [1.3.1-2build1]  python-yaml{a} [3.11-3build1]  
  python2.7{a} [2.7.12-1ubuntu0~16.04.12]  python2.7-minimal{a} [2.7.12-1ubuntu0~16.04.12]  python3{a} [3.5.1-3]  python3-minimal{a} [3.5.1-3]  python3-psutil{a} [3.4.2-1ubuntu0.1]  python3-pyudev{a} [0.16.1-3]  
  python3-yaml{a} [3.11-3build1]  python3.5{a} [3.5.2-2ubuntu0~16.04.11]  python3.5-minimal{a} [3.5.2-2ubuntu0~16.04.11]  realpath{a} [8.25-2ubuntu3~16.04]  scylla-enterprise [2019.1.11-0.20200810.966418404-0ubuntu1~xenial]  
  scylla-enterprise-conf{a} [2019.1.12-0.20200729.fc61a884b-0ubuntu1~xenial]  scylla-enterprise-jmx{a} [2019.1.11-20200727.690bbe5-0ubuntu1~xenial]  scylla-enterprise-kernel-conf{a} [2019.1.11-0.20200727.c98c0150d-0ubuntu1~xenial]  
  scylla-enterprise-server{a} [2019.1.11-0.20200727.c98c0150d-0ubuntu1~xenial]  scylla-enterprise-tools{a} [2019.1.11-20200727.22cc383b70-0ubuntu1~xenial]  
  scylla-enterprise-tools-core{a} [2019.1.11-20200727.22cc383b70-0ubuntu1~xenial]  scylla-gcc73-libatomic1{a} [7.3.0-3ubuntu2~xenialppa1]  scylla-gcc73-libstdc++6{a} [7.3.0-3ubuntu2~xenialppa1]  sgml-base{a} [1.26+nmu4ubuntu1]  
  ucf{a} [3.0036]  x11-common{a} [1:7.7+13ubuntu3.1]  xml-core{a} [0.13+nmu2]  
0 packages upgraded, 103 newly installed, 0 to remove and 2 not upgraded.
Need to get 111 MB of archives. After unpacking 363 MB will be used.
```

Another solution can involve a reverse dependency, I don't know if it will work but:
maybe we can have another meta-package that all of the other pakcages depends on (scylla-enterprise-reverse-2019.1.X), all
other packages that are 2019.1.X will be dependent uppon.
The if we want to install version 2019.1.X we first pin this package with the specific version and only then attempt to install
the scylla-enterprise package, maybe then apt will have an "easier" choice in regards to which pakcage should it choose.
I will try this in the near future and report, but the above solution works and maybe we should consider adding the dependency 
ranges regardless, this will prevent an accidental installation the non requested scylla version.

**Comment by avikivity** [comment]
I think we should implement @syuu1228 's idea. We need it anyway for master.

**Comment by eliransin** [comment]
> In such condition, I think specifying version for each packages is simplest way (we use it for downgrade).
> Something like:
> 
> ```
> sudo apt-get install scylla=3.x.y\* scylla-server=3.x.y\* scylla-jmx=3.x.y\* scylla-tools=3.x.y\* scylla-tools-core=3.x.y\* scylla-kernel-conf=3.x.y\* scylla-conf=3.x.y\*
> ```

@avikivity are you refering to this?
If you do, what is there to implement except specifying this line somewhere? (in the docs or somewhere else)
I would still add the dependency part since without it users will be mislead that they are installing a certain version while
in fact they are installing the latest one.

**Comment by avikivity** [comment]
No, the idea of having the same full version for all packages.

https://github.com/scylladb/scylla/issues/5514#issuecomment-576516050

**Comment by avikivity** [comment]
But I see he later said it doesn't work. If debian doesn't support it, we don't need to support it either.

**Comment by eliransin** [comment]
OK, the idea that will definitely work is the one from @syuu1228 comment, simply move to a single package.
However the real bug here is that a user that tries to install an older version end up with a newer one without him knowing,
IMO it is better if the installation will fail if the user tries to do this and not end up with a version he doesn't expect.

**Comment by avikivity** [comment]
Single package is too much. We want the ability not to install jmx or kernel-conf.

**Comment by slivne** [comment]
@syuu1228 / @penberg ping

**Comment by eliransin** [comment]
@avikivity - so we are left with 2 options:
1. Close this with the "wontfix" label.
2. We have the extreme solution of making a unified single debian package (which now with submoduling all of other 
dependencies is doable with a reasonable effort.
I just need to know if there is a reason to keep pushing this.

**Comment by avikivity** [comment]
There is now (with the submodules) a third option.

3. add dependency on the exact version between scylla-x.y.z and scylla-tools-java-x.y.z (and other subpackages).

We could not do this before[[1]](https://github.com/scylladb/scylla/issues/5514#issuecomment-576516050) because subpackages had different git hashes in their names.

This won't automatically pull an older version of scylla-tools-java, but it will error out instead of installing a mixed version. The user can then correct by asking apt to install and older version of scylla and all its subpackages.

Your option 1 is a bit too much, we want to be able to install scylla-tools-java on machines that don't have scylla-server, and we want to be able to get rid of scylla-jmx for installations that don't need it.

@syuu1228 is this option 3 doable or did I miss something?

**Comment by avikivity** [comment]
Not a regression, and depends on recent changes in submodules. Not backporting.
